spacy/tests/pipeline/test_entity_linker.py

```diff
@@ -502,20 +503,20 @@ def create_kb(vocab):
     assert doc[2].ent_kb_id_ == "Q2"
 
     def get_lowercased_candidates(kb, span):
-        return kb.get_alias_candidates(span.text.lower())
+        return kb._get_alias_candidates(span.text.lower())
 
     def get_lowercased_candidates_batch(kb, spans):
         return [get_lowercased_candidates(kb, span) for span in spans]
 
     @registry.misc("spacy.LowercaseCandidateGenerator.v1")
     def create_candidates() -> Callable[
-        [InMemoryLookupKB, "Span"], Iterable[Candidate]
+        [InMemoryLookupKB, "Span"], Iterable[InMemoryCandidate]
```

Comment:
Why do we need to change the type here to `InMemoryCandidate`? Does it really matter what the exact type of the candidate is, can't the code be more generic to deal with any kind?

Reply:
(same question for the next one)

Reply:
We don't need to. By default I try to be as specific as possible when it comes to typing. If you prefer, we can stick with `Iterable[Candidate]` here, I don't mind.

Reply:
Basically my thinking was that all candidate generators would have the same return type, and thus be more uniform. But I can also see the reasoning why to make it more specific if we know that that's the exact subtype that'll come up. I'm not sure which is best, tbh.

Reply:
Basically my thinking was that all candidate generators would have the same return type, and thus be more uniform. But I can also see the reasoning why to make it more specific if we know that that's the exact subtype that'll come up. I'm not sure which is best, tbh.

Reply:
Right, so my approach is usally: when writing interfaces/classes/functions, i. e. things that will be used in multiple places, go with the most abstract type supporting the operations in there. When writing "consumer" code, i. e. you are writing something that won't be reused anywhere else, use the most specific types possible as this makes reading the code easier and supports better auto-complete.

This is my two cents. Having said that, I don't care much in this case anyway, as it's just a test.

Reply:
Yea, to me it was more of a question also how we type the actual default candidate generator.

In my opinion, we don't get any added advantages of typing the more specific type, because we wouldn't want any other code to rely on that specific subtype, because at that point we can't switch in a different candidate generator.

Reply:
Isn't developers' knowledge about the returned type an advantage in itself? Typing is also a form of documentation and it would be more understandable if we type the specific type that is returned by the default candidate retrieval function IMO.

I defer to your judgement here though, this is not important enough for a debate :grin:

Reply:
But the developer shouldn't even really need to know about the returned type other than it being a `Candidate`, if we want the candidate generators to be uniform in their output, you know?

So yea I do think it's best practice throughout the code base to have architectures that are meant to be swappable with eachother, to have the exact same output type, even if in reality a subtype is returned.

Reply:
That's a good point.